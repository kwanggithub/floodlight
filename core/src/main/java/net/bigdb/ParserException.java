package net.bigdb;

import java.util.List;

import net.bigdb.BigDBException;

public class ParserException extends BigDBException {

    private static final long serialVersionUID = 6591331000339084713L;

    private List<String> errors;

    public ParserException(String message, List<String> parseErrors) {
        super(message);
        errors = parseErrors;
    }

    public List<String> getErrors() {
        return errors;
    }

    @Override
    public String getMessage() {
        StringBuilder sb = new StringBuilder();
        sb.append(super.getMessage());
        sb.append(";");
        sb.append(newLine);
        if (errors != null) {
            for (String s : this.errors) {
                sb.append(s);
                sb.append(";");
                sb.append(newLine);
            }
        }
        return sb.toString();
    }

}
