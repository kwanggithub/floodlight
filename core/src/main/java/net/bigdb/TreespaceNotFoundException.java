package net.bigdb;

public class TreespaceNotFoundException extends BigDBException {
    private static final long serialVersionUID = 1L;

    public TreespaceNotFoundException(String treespace) {
        super("Treespace "+treespace+" not found", Type.NOT_FOUND);
    }
}
