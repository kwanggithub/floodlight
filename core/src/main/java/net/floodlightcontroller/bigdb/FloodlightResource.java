package net.floodlightcontroller.bigdb;

import net.floodlightcontroller.core.IFloodlightProviderService;
import net.floodlightcontroller.core.module.FloodlightModuleContext;
import net.floodlightcontroller.counter.ICounterStoreService;
import net.floodlightcontroller.device.IDeviceService;

public class FloodlightResource {
    
    // Only allow direct access to the data member (and the setter) from the
    // other classes in the package (e.g. the BigDBService implementation),
    // not the subclasses.
    // private static IFloodlightProviderService floodlightProvider;
    
    private static FloodlightModuleContext moduleContext;
    
    public static IDeviceService getDeviceService() {
        return moduleContext.getServiceImpl(IDeviceService.class);
    }
    
    public static IFloodlightProviderService getFloodlightProvider() {
        return moduleContext.getServiceImpl(IFloodlightProviderService.class);
    }
    
    public static ICounterStoreService getCounterService() {
        return moduleContext.getServiceImpl(ICounterStoreService.class);
    }
    
    public static FloodlightModuleContext getModuleContext() {
        return moduleContext;
    }
    
    static void setModuleContext(FloodlightModuleContext context) {
        FloodlightResource.moduleContext = context;
    }
}
