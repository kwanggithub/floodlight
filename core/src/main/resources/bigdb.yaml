treespaces:
  - name: controller
    data_sources:
      - name: config
        config: true
        implementation_class: net.bigdb.data.memory.MemoryDataSource
      - name: node-local-config
        config: true
        implementation_class: net.bigdb.data.memory.MemoryDataSource
treespaces:
  - name: controller
    data_sources:
        - name: config
          config: true
          implementation_class: net.bigdb.data.memory.MemoryDataSource
    data_source_mappings:
      - data_source: config
        predicate: Config
      - data_source: "$data-source"
        predicate: "!Config"
#   Need to figure out where to put the schema files for production.
#   May use module properties.
    module_search_paths:
      - path: src/main/resources/schema
        recursive: true
    modules:
      - name: floodlight
        revision: 2012-10-22
      - name: topology
      - name: aaa
      - name: os
        revision: 2013-06-12
      - name: cluster
        revision: 2013-06-12
      - name: tag
      - name: device-security
        revision: 2013-07-11


