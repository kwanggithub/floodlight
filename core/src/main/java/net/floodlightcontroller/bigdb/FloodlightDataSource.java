package net.floodlightcontroller.bigdb;

import net.bigdb.BigDBException;
import net.bigdb.data.ServerDataSource;
import net.bigdb.schema.Schema;
import net.floodlightcontroller.core.module.FloodlightModuleContext;

public class FloodlightDataSource extends ServerDataSource {

    public FloodlightDataSource(FloodlightModuleContext context, Schema schema)
            throws BigDBException  {
        super("controller-data-source", schema);
        FloodlightResource.setModuleContext(context);
    }
}
