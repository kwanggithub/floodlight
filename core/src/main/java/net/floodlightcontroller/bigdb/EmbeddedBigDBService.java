package net.floodlightcontroller.bigdb;

import java.io.File;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import net.bigdb.BigDBException;
import net.bigdb.TreespaceNotFoundException;
import net.bigdb.auth.AuthConfig;
import net.bigdb.auth.session.SimpleSessionManager;
import net.bigdb.config.DataSourceConfig;
import net.bigdb.config.DataSourceMappingConfig;
import net.bigdb.config.ModuleConfig;
import net.bigdb.config.RootConfig;
import net.bigdb.config.TreespaceConfig;
import net.bigdb.data.ServerDataSource;
import net.bigdb.service.Service;
import net.bigdb.service.Treespace;
import net.bigdb.service.internal.ServiceImpl;
import net.floodlightcontroller.core.module.FloodlightModuleContext;
import net.floodlightcontroller.core.module.FloodlightModuleException;
import net.floodlightcontroller.core.module.IFloodlightModule;
import net.floodlightcontroller.core.module.IFloodlightService;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.common.collect.Lists;

public class EmbeddedBigDBService implements IFloodlightModule, IBigDBService {

    public static final String CONTROLLER_TREESPACE_NAME = "controller";

    protected final static Logger logger =
            LoggerFactory.getLogger(EmbeddedBigDBService.class);
    protected final ServiceImpl bigdbService;
    private ServerDataSource floodlightDataSource;
    private final String configFile;

    protected AuthConfig authConfig;

    public EmbeddedBigDBService() throws BigDBException  {
        this(null);
    }

    public EmbeddedBigDBService(String configFile) throws BigDBException  {
        this.configFile = configFile;
        this.bigdbService = new net.bigdb.service.internal.ServiceImpl();
    }

    public static RootConfig getDefaultConfig(List<ModuleConfig> moduleConfig,
                                              AuthConfig authConfig) {
        RootConfig rc = new RootConfig();
        TreespaceConfig tc = new TreespaceConfig();
        tc.name = "controller";
        tc.modules = moduleConfig;

        DataSourceConfig dcc = new DataSourceConfig();
        dcc.implementation_class = "net.bigdb.data.memory.MemoryDataSource";
        dcc.name = "config";
        dcc.config = true;
        tc.data_sources = Lists.newArrayList(dcc);
        
        DataSourceMappingConfig dsmc1 = new DataSourceMappingConfig();
        dsmc1.data_source = "config";
        dsmc1.predicate = "Config";
        DataSourceMappingConfig dsmc2 = new DataSourceMappingConfig();
        dsmc2.data_source = "$data-source";
        dsmc2.predicate = "!Config";
        
        tc.data_source_mappings = Lists.newArrayList(dsmc1, dsmc2);
        rc.treespaces = Lists.newArrayList(tc);
        
        if (authConfig == null) {
            authConfig = new AuthConfig().setParam(AuthConfig.SESSION_MANAGER, 
                                                   SimpleSessionManager.class);
            authConfig.setParam(AuthConfig.ENABLE_NULL_AUTHENTICATION, true);
        }
        return rc;
    }
    
    @Override
    public Collection<Class<? extends IFloodlightService>> getModuleServices() {
        Collection<Class<? extends IFloodlightService>> l =
                new ArrayList<Class<? extends IFloodlightService>>();
        l.add(IBigDBService.class);
        return l;
    }

    @Override
    public Map<Class<? extends IFloodlightService>, IFloodlightService>
            getServiceImpls() {
        Map<Class<? extends IFloodlightService>, IFloodlightService> m =
                new HashMap<Class<? extends IFloodlightService>,
                IFloodlightService>();
        // We are the class that implements the service
        m.put(IBigDBService.class, this);
        return m;
    }

    @Override
    public Collection<Class<? extends IFloodlightService>> getModuleDependencies() {
        return null;
    }

    @Override
    public void init(FloodlightModuleContext context)
            throws FloodlightModuleException {
        try {
            Map<String, String> configParams = context.getConfigParams(this);
            String bigDBConfigFileName = configParams.get("BigDBConfigFileName");

            if(this.authConfig != null)
                bigdbService.setAuthConfig(authConfig);

            if (bigDBConfigFileName != null) {
                File bigdbConfigFile = new File(bigDBConfigFileName);
                bigdbService.initializeFromFile(bigdbConfigFile);
            } else {
                // No config file was specified and/or the floodlight config
                // directory could not be resolved, so use the default config
                // file that comes from a resource.
                bigdbService.initializeFromResource(this.configFile);
            }
            completeInit(context);
        } catch (Exception e) {
            logger.error("Failed to start bigDB service: " + e.getMessage(), e);
            // re-throw the exception
            throw new FloodlightModuleException(e.getMessage(), e);
        }
    }
    
    protected void completeInit(FloodlightModuleContext context) 
            throws Exception {
        Treespace treespace = bigdbService.getTreespace("controller");
        if (treespace == null)
            throw new Exception("Controller treespace not found");
        floodlightDataSource = new FloodlightDataSource(context, 
                                                        treespace.getSchema());
        floodlightDataSource.setTreespace(treespace);
        treespace.registerDataSource(floodlightDataSource);
    }

    public void setAuthConfig(AuthConfig authConfig) {
        this.authConfig = authConfig;
    }

    @Override
    public void startUp(FloodlightModuleContext context) {
        try {
            bigdbService.startup();
        } catch (BigDBException e) {
            throw new RuntimeException("BigDBService failed to startup: " + e.getMessage(), e);
        }
    }

    @Override
    public Service getService() {
        return bigdbService;
    }

    @Override
    public Treespace getControllerTreespace() throws TreespaceNotFoundException {
        return bigdbService.getTreespace(CONTROLLER_TREESPACE_NAME);
    }

    @Override
    public ServerDataSource getControllerDataSource() {
        return floodlightDataSource;
    }

    @Override
    public void run() throws BigDBException {
        try {
            bigdbService.run();
        } catch (BigDBException e) {
            logger.error("Failed to start bigdb rest service: " + e.toString());
            throw e;
        }
    }

    @Override
    public void stop() throws BigDBException {
        this.bigdbService.stop();
    }
}
