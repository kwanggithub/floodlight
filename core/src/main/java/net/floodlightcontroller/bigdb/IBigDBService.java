package net.floodlightcontroller.bigdb;

import net.bigdb.BigDBException;
import net.bigdb.TreespaceNotFoundException;
import net.bigdb.data.ServerDataSource;
import net.bigdb.service.Service;
import net.bigdb.service.Treespace;
import net.floodlightcontroller.core.module.IFloodlightService;

/**
 * This is the interface through which floodlight modules access BigDB.
 * They can access the BigDB Service interface and use it to read or write
 * data to BigDB or call attachDataSource to set up custom data sources.
 *
 * @author rob.vaterlaus@bigswitch.com
 */
public interface IBigDBService extends IFloodlightService {
    /**
     * Access the BigDB service interface.
     * @return
     */
    public Service getService();

    /**
     * Access the main BigDB treespace for the controller.
     * @return
     * @throws TreespaceNotFoundException
     */
    public Treespace getControllerTreespace() throws TreespaceNotFoundException;

    /**
     * Access the data source for the controller for adding providers
     * of dynamic state to BigDB.
     * @return the BigDB request router for the controller data source
     */
    public ServerDataSource getControllerDataSource();

    /**
     * Begin running the BigDB service. This should only be called from the
     * core floodlight code.
     * @throws BigDBException
     */
    public void run() throws BigDBException;

    public void stop() throws BigDBException;
}
