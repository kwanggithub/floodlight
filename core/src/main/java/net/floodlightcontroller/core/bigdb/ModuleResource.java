package net.floodlightcontroller.core.bigdb;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

import net.bigdb.BigDBException;
import net.bigdb.data.annotation.BigDBProperty;
import net.bigdb.data.annotation.BigDBQuery;
import net.floodlightcontroller.bigdb.FloodlightResource;
import net.floodlightcontroller.core.module.FloodlightModuleLoader;
import net.floodlightcontroller.core.module.IFloodlightModule;
import net.floodlightcontroller.core.module.IFloodlightService;

public class ModuleResource extends FloodlightResource {

    private final FloodlightModuleLoader moduleLoader;

    public ModuleResource(FloodlightModuleLoader moduleLoader) {
        this.moduleLoader = moduleLoader;
    }

    @SuppressFBWarnings(value="EQ_COMPARETO_USE_OBJECT_EQUALS")
    // Note: this class has a natural ordering that is inconsistent with equals.
    public static class ModuleInfo implements Comparable<ModuleInfo> {
        String moduleName;
        boolean loaded;
        Set<String> dependencies;
        Set<String> services;

        @BigDBProperty(value="module-name")
        public String getModuleName() {
            return moduleName;
        }
        public void setModuleName(String moduleName) {
            this.moduleName = moduleName;
        }
        public boolean getLoaded() {
            return loaded;
        }
        public void setLoaded(boolean loaded) {
            this.loaded = loaded;
        }
        public Set<String> getDependencies() {
            return dependencies;
        }
        public void setDependencies(Set<String> dependencies) {
            this.dependencies = dependencies;
        }
        public Set<String> getServices() {
            return services;
        }
        public void setServices(Set<String> services) {
            this.services = services;
        }
        @Override
        public int compareTo(ModuleInfo o) {
            return moduleName.compareTo(o.moduleName);
        }
    }

    /**
     * Retrieves all modules and their dependencies available to Floodlight.
     * @return Information about modules available or loaded.
     */
    @BigDBQuery
    public List<ModuleInfo> getAllModules() throws BigDBException {
        List<ModuleInfo> moduleList = new ArrayList<ModuleInfo>();
        Set<String> loadedModules = new HashSet<String>();

        for (IFloodlightModule m : moduleLoader.getLoadedModuleSet()) {
            loadedModules.add(m.getClass().getCanonicalName());
        }

        for (String moduleName : moduleLoader.getModuleNameMap().keySet() ) {
            ModuleInfo moduleInfo = new ModuleInfo();

            IFloodlightModule module =
                    moduleLoader.getModuleNameMap().get(
                                moduleName);

            moduleInfo.setModuleName(module.getClass().getCanonicalName());

            Collection<Class<? extends IFloodlightService>> deps =
                    module.getModuleDependencies();
            if (deps == null)
                deps = new HashSet<Class<? extends IFloodlightService>>();
            Map<String,Object> depsMap = new HashMap<String, Object> ();
            for (Class<? extends IFloodlightService> service : deps) {
                Object serviceImpl = getModuleContext().getServiceImpl(service);
                if (serviceImpl != null)
                    depsMap.put(service.getCanonicalName(), serviceImpl.getClass().getCanonicalName());
                else
                    depsMap.put(service.getCanonicalName(), "<unresolved>");

            }
            moduleInfo.setDependencies(depsMap.keySet());
            
            Collection<Class<? extends IFloodlightService>> provides = 
                    module.getModuleServices();
            if (provides == null)
                provides = new HashSet<Class<? extends IFloodlightService>>();
            Map<String,Object> providesMap = new HashMap<String,Object>();
            for (Class<? extends IFloodlightService> service : provides) {
                providesMap.put(service.getCanonicalName(), module.getServiceImpls().get(service).getClass().getCanonicalName());
            }
            moduleInfo.setServices(providesMap.keySet());

            boolean loaded = false;
            // check if this module is loaded directly
            if (loadedModules.contains(module.getClass().getCanonicalName())) {
                loaded = true;          
            } else {
                // if not, then maybe one of the services it exports is loaded
                for (Class<? extends IFloodlightService> service : provides) {
                    String modString = module.getServiceImpls().get(service).getClass().getCanonicalName();
                    if (loadedModules.contains(modString))
                        loaded = true;
                }
            }
            moduleInfo.setLoaded(loaded);
            
            moduleList.add(moduleInfo);
        }            
        Collections.sort(moduleList);
        
        return moduleList;
    }
}
