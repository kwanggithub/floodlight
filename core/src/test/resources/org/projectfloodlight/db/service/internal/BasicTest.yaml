treespaces:
  - name: BasicTest
    data_sources:
        - name: config
          implementation_class: org.projectfloodlight.db.data.memory.MemoryDataSource
    data_source_mappings:
      - data_source: config
        predicate: Config
      - data_source: "$data-source"
        predicate: "!Config"
    module_search_paths:
      - path: src/test/resources/org/projectfloodlight/db/service/internal
        recursive: true
    modules:
      - name: BasicTest
auth_config:
    enable_null_authentication: true
    default_authorization_hook: org.projectfloodlight.db.auth.NullAuthorizationHook
    session_manager: org.projectfloodlight.db.auth.session.SimpleSessionManager
