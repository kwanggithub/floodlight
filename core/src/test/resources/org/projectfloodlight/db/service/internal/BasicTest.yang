module BasicTest {
    namespace org.projectfloodlight.db;
    prefix "ct";
    
    import bigdb {
        prefix bigdb;
    }

    description "Basic test schema";
    
    organization "BigDB Inc.";

    typedef priority-type {
        type int32;
        default 1000;
    }

    // Somewhat artificially factoring these fields into a grouping. This is
    // as a test for a data source list corruption bug. See the comment below
    // for the person-dummy-state node.
    grouping name-fields {
        leaf first-name {
            bigdb:column-header "First Name";
            type string;
        }
        
        leaf last-name {
            bigdb:column-header "Last Name";
            type string;
        }
    }

    list person {

        key id;
        
        leaf id {
            type string;
        }
        
        uses ct:name-fields;

        leaf-list nickname {
            type string;
        }
        
        list job {
            key "employer-id employee-id";
            
            min-elements 0;
            max-elements 4;
            
            leaf employer-id {
                type string;
            }
            leaf employee-id {
                type string;
            }
            leaf title {
                type string;
            }
        }
        
        list address {

            key "street city state zip-code";
            leaf street {
                type string;
            }
            
            leaf city {
                type string;
            }
            
            leaf state {
                type string;
            }
            
            leaf zip-code {
                type string;
            }
        }
    }
    
    // This list is here to test for a bug where the data source list for
    // nodes from a grouping would get corrupted if the grouping was used for
    // both config and operational state. For the bug to occur it's necessary
    // for the name of this node to be alphabetically after the "person" node.
    // There isn't a specific @Test for this bug. If the bug occurs pretty
    // much all of the other tests fail. So none of the tests actually access
    // this node, but it should be left here as a regression test for the
    // data source list corruption bug.
    list person-dummy-state {
        config false;
        bigdb:data-source "controller-data-source";
        uses ct:name-fields;
    }

    list group {
        key "name";
        leaf name {
            type string;
        }
    }

    list external-group {
        key "name";
        leaf name {
            type string;
        }
        list external-user {
            key "company-id user-id";
            leaf company-id {
                type string;
            }
            leaf user-id {
                bigdb:case-sensitive false;
                type string;
            }
        }
    }

    leaf allow-empty-leaf {
        bigdb:allow-empty-string true;
        type string {
            pattern "abc.*";
        }
    }

    leaf allow-empty-type {
        type string {
            bigdb:allow-empty-string true;
            pattern "abc.*";
        }
    }

    list prioritized-list {
        key "priority";
        leaf priority {
            type uint32;
        }
        leaf name {
            type string;
        }
    }

    list update-test {
        key "child1";
        leaf child1 {
            type string;
        }
        container child2 {
            leaf child21 {
                type string;
            }
            leaf child22 {
                type int32;
            }
        }
        leaf child3 {
            type string;
        }
    }

    container default-values {
        leaf leaf-default {
            type string;
            default "foobar";
        }
        leaf typedef-default {
            type priority-type;
        }
        leaf typedef-default-override {
            type priority-type;
            default 2000;
        }
    }

    container empty-container {
        container child1 {
            container child12 {
                leaf child121 {
                    type int32;
                }
            }
        }
        container child2 {
            container child21 {
                leaf child211 {
                    type string;
                    default "foo";
                }
            }
        }
        list child3 {
            key child31;
            leaf child31 {
                type string;
            }
            container child32 {
                leaf child321 {
                    type int32;
                }
            }
            container child33 {
                leaf child331 {
                    type string;
                    default "bar";
                }
            }
        }
     }
}
