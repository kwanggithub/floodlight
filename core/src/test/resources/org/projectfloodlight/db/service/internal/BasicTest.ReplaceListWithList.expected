[ {
  "first-name" : "Steve",
  "id" : "555-66-7777",
  "last-name" : "Abbott"
}, {
  "first-name" : "Ron",
  "id" : "666-77-8888",
  "last-name" : "Bell"
} ]